### PR TITLE
Migrate to Jetpack Compose

### PR DESCRIPTION
**Pull request by ArnyminerZ**
Migrated to Jetpack Compose and added some testing for the activity rejection and accept.

### CODE REVIEW

**Comment by ArnyminerZ** [comment]
@rfc2822 This should be ready. Take your time, I know it may be a lot to process.

When using Jetpack Compose trailing commas really help the linter to correctly format, that's why I've used them. I've tried to use them only on Compose functions.

There's now a new class called `ParcelableColorScheme` with a matching extra in `TrustCertificateActivity.EXTRA_COLOR_SCHEME`. This class is for passing the color theme of the source application to here. Since every application that uses cert4android, and makes use of Jetpack Compose will have its own theme, it can be passed along with:
```kotlin
val intent = Intent(context, TrustCertificateActivity::class.java)
intent.putExtra(TrustCertificateActivity.EXTRA_CERTIFICATE, MaterialTheme.colorScheme.parcelize())
startActivity(intent)
```
_Note that `MaterialTheme` is a predefined object, which matches the currently defined color scheme._

I've also added a new test that launches the app, and tests that it works correctly. This is that if:
- A: The user rejects the certificate, check that it's still not trusted after closing the activity.
- B: The user wants to accept the certificate, check that the checkbox has to be marked before clicking the trust button; and that clicking the button does trust the certificate.

**Comment by ArnyminerZ** [comment]
Tests are failing here but not on my computer... I really don't know what else to try 😅

**Comment by rfc2822** [comment]
First, thanks for your efforts! I'm not happy that changing to Compose adds +600 lines, but maybe it's not possible otherwise.

> There's now a new class called ParcelableColorScheme with a matching extra in TrustCertificateActivity.EXTRA_COLOR_SCHEME. This class is for passing the color theme of the source application to here. Since every application that uses cert4android, and makes use of Jetpack Compose will have its own theme, it can be passed along with:

Ah, I see, the theme is passed explicitly instead of implicitly over XML resources.

I have a suggestion to save some code (and complexity): Now the approach is to pass the theme to every call of the Activity, so the theme must be Parcelable. However we don't need every Activity to be themed differently, because the calling app will have only one (active) theme and this one could be set for cert4android too.

**So I suggest to have some global initializing method that stores the theme from the calling app into cert4android and that doesn't need to be passed around.**

For instance there could be a `theme` variable (default value: some default Material theme without special colors) in cert4android.`Constants` that must be set from the calling app before using cert4android. (I'd also rename `Constants` to `Cert4android`.)

Then we don't need the `ParcelableColorScheme` and probably also the default theme, which should make the whole thing more compact.

What do you think?

**Comment by ArnyminerZ** [comment]
Yeah, that was my first thought. I think it will work better than always passing the color scheme. I don't exactly know how much it weights, but it can also cause some problems. So I will change to an independent variable.

**Comment by ArnyminerZ** [comment]
I've now added `Cert4AndroidTheme`, which provides the current theme instead of the extra. So the apps that use cert4android just have to call `Cert4AndroidTheme.setColorScheme` on the `onCreate` method of the `Application`, for example, and the theme will be updated accordingly.

**Comment by rfc2822** [comment]
> I've now added `Cert4AndroidTheme`, which provides the current theme instead of the extra. So the apps that use cert4android just have to call `Cert4AndroidTheme.setColorScheme` on the `onCreate` method of the `Application`, for example, and the theme will be updated accordingly.

Can we drop the generated theme (160 lines) then? And add to documentation that the theme must be set before using cert4android.

File: src/main/res/values/styles.xml
Comment (rfc2822) [inline src/main/res/values/styles.xml:1]: If it's not needed, please remove it

**Comment by rfc2822** [comment]
> After researching a bit around that failing test. Looks like it's a problem with Github actions. I hope they fix it soon... Otherwise we can just disable those tests on Actions, and wait until a fix is found.

The failing tests seem to be failing Android tests. How are they related to Github Actions? I think there must be some other problem. Maybe disabled animations or something like that…

**Comment by rfc2822** [comment]
We had so many problems with automated UI tests, I recommend not using them unless we have the capacity to maintain them, which is much more work than it seems. I'll remove the for now and have a look at the other things :)

**Comment by rfc2822** [comment]
I have reordered the dependencies (please always list them in the same way as in the DAVx5 repo, because I have my routines where I update the dependencies and it should everywhere look the same ;)).

We probably really need the compose preview dependency, too. I didn't know that it's only for the debug build, so it shouldn't have impact.

However the theme problem is not solved yet – when cert4android should be a library for any Compose app, it can't just use `MdcTheme`, because the calling application may not use a Material XML theme but for instance a native Compose Material3 theme.

So we probably need to pass the theme somehow, and use some unstyled default theme for the previews / when no theme is passed. Can we pass the theme somehow?

**Comment by ArnyminerZ** [comment]
Then I will just recover the old `Cert4AndroidTheme.getColorScheme` and `Cert4AndroidTheme.setColorScheme` approach, I think it should work fine.

**Comment by rfc2822** [comment]
It doesn't work with the current DAVx5, which shows at compile time:

> at.bitfire.davdroid.app-mergeStandardDebugResources-85:/layout/delete_collection.xml:23: error: resource style/TextView.Heading (aka at.bitfire.davdroid:style/TextView.Heading) not found.
error: failed linking file resources.

Can you check/reproduce that?

Maybe because cert4android uses M3, while DAVx5 uses M2? I think we shouldn't mix in M3 yet.

**Comment by rfc2822** [comment]
I have created a new DAVx5 branch that works with this cert4android PR: https://github.com/bitfireAT/davx5/tree/cert4android-compose

I think basically its finished although there are a few questions:

- Updating Kotlin / Compose is now only possible when cert4android and DAVx5 are updated at the same time, right? Otherwise I guess the Kotlin compiler versions collide.
- Would it maybe make sense to outsource this to a "real" library instead of a submodule? (In a separate PR) Then we could have a sample Activity in the separate project, the lib in the separate project and because it would be pre-compiled we wouldn't need to care about the Compose/Kotlin version
- I had to modify the text styling and it's now really ugly ;) Can you have another look @ArnyminerZ ?

**Comment by ArnyminerZ** [comment]
Okay, so now is virtually identical. Here the old one for reference:
cert4android-old-1
cert4android-old-2

And the new one (with Jetpack Compose):
cert4android-4

**Comment by ArnyminerZ** [comment]
I've now created #16 to import all the changes

**Pull request closed without merging**